src/cider/type/plan/FunctionExpr.cpp

```diff
@@ -0,0 +1,307 @@
+    return false;
+  }
+  const auto rhs_func_oper =
+      dynamic_cast<const FunctionOperWithCustomTypeHandling*>(&rhs);
+  if (!rhs_func_oper) {
+    return false;
+  }
+  if (getName() != rhs_func_oper->getName()) {
+    return false;
+  }
+  if (getArity() != rhs_func_oper->getArity()) {
+    return false;
+  }
+  for (size_t i = 0; i < getArity(); ++i) {
+    if (!(*getArg(i) == *(rhs_func_oper->getArg(i)))) {
+      return false;
+    }
+  }
+  return true;
+}
+
+JITTypeTag ext_arg_type_to_JIT_type_tag(const ExtArgumentType ext_arg_type) {
+  switch (ext_arg_type) {
+    case ExtArgumentType::Bool:  // pass thru to Int8
+    case ExtArgumentType::Int8:
+      return JITTypeTag::INT8;
+    case ExtArgumentType::Int16:
+      return JITTypeTag::INT16;
+    case ExtArgumentType::Int32:
+      return JITTypeTag::INT32;
+    case ExtArgumentType::Int64:
+      return JITTypeTag::INT64;
+    case ExtArgumentType::Float:
+      return JITTypeTag::FLOAT;
+    case ExtArgumentType::Double:
+      return JITTypeTag::DOUBLE;
+    case ExtArgumentType::ArrayInt64:
+    case ExtArgumentType::ArrayInt32:
+    case ExtArgumentType::ArrayInt16:
+    case ExtArgumentType::ArrayBool:
+    case ExtArgumentType::ArrayInt8:
+    case ExtArgumentType::ArrayDouble:
+    case ExtArgumentType::ArrayFloat:
+    case ExtArgumentType::ColumnInt64:
+    case ExtArgumentType::ColumnInt32:
+    case ExtArgumentType::ColumnInt16:
+    case ExtArgumentType::ColumnBool:
+    case ExtArgumentType::ColumnInt8:
+    case ExtArgumentType::ColumnDouble:
+    case ExtArgumentType::ColumnFloat:
+    case ExtArgumentType::TextEncodingNone:
+    case ExtArgumentType::ColumnListInt64:
+    case ExtArgumentType::ColumnListInt32:
+    case ExtArgumentType::ColumnListInt16:
+    case ExtArgumentType::ColumnListBool:
+    case ExtArgumentType::ColumnListInt8:
+    case ExtArgumentType::ColumnListDouble:
+    case ExtArgumentType::ColumnListFloat:
+      return JITTypeTag::VOID;
+    default:
+      CHECK(false);
+  }
+  CHECK(false);
+  return JITTypeTag::INVALID;
+}
+
+bool ext_func_call_requires_nullcheck(const Analyzer::FunctionOper* function_oper) {
+  const auto& func_ti = function_oper->get_type_info();
+  for (size_t i = 0; i < function_oper->getArity(); ++i) {
+    const auto arg = function_oper->getArg(i);
+    const auto& arg_ti = arg->get_type_info();
+    if ((func_ti.is_array() && arg_ti.is_array()) ||
+        (func_ti.is_bytes() && arg_ti.is_bytes())) {
+      // If the function returns an array and any of the arguments are arrays, allow NULL
+      // scalars.
+      // TODO: Make this a property of the FunctionOper following `RETURN NULL ON NULL`
+      // semantics.
+      return false;
+    } else if (!arg_ti.get_notnull() && !arg_ti.is_buffer()) {
+      return true;
+    } else {
+      continue;
+    }
+  }
+  return false;
+}
+
+inline SQLTypeInfo get_sql_type_from_JIT_type_tag(const JITTypeTag jit_type) {
+  switch (jit_type) {
```

Comment:
better to define in typeUtils.h to reuse code?

Reply:
can't do that. It's not proper to include `function/ExtensionFunctionsWhitelist.h` in `TypeUtils.cpp` and it also will cause segfault.